src/store_client.cc

```diff
@@ -153,59 +155,58 @@ storeClientListAdd(StoreEntry * e, void *data)
     return sc;
 }
 
-void
-store_client::callback(ssize_t sz, bool error)
+static void
+CallbackClientSide(store_client *sc)
 {
-    size_t bSz = 0;
+    sc->callbackClientSide();
+}
 
-    if (sz >= 0 && !error)
-        bSz = sz;
+void
+store_client::callbackClientSide()
+{
+    assert(_callback.pending());
 
-    StoreIOBuffer result(bSz, 0,copyInto.data);
+    clientSideCaller = nullptr;
+    StoreIOBuffer result(copiedSize, copyInto.offset, copyInto.data);
+    result.flags.error = copyInto.flags.error;
 
-    if (sz < 0 || error)
-        result.flags.error = 1;
-
-    result.offset = cmp_offset;
-    assert(_callback.pending());
-    cmp_offset = copyInto.offset + bSz;
-    STCB *temphandler = _callback.callback_handler;
+    auto temphandler = _callback.callback_handler;
     void *cbdata = _callback.callback_data;
-    _callback = Callback(NULL, NULL);
-    copyInto.data = NULL;
+    _callback = store_client::Callback(nullptr, nullptr);
+    copyInto.data = nullptr;
+    copiedSize = 0;
 
     if (cbdataReferenceValid(cbdata))
         temphandler(cbdata, result);
 
     cbdataReferenceDone(cbdata);
 }
 
-static void
-storeClientCopyEvent(void *data)
+void
+store_client::callback(ssize_t sz, bool error)
 {
-    store_client *sc = (store_client *)data;
-    debugs(90, 3, "storeClientCopyEvent: Running");
-    assert (sc->flags.copy_event_pending);
-    sc->flags.copy_event_pending = false;
+    copiedSize = (sz > 0 && !error) ? sz : 0;
```

Comment:
if store_client::callback() is called several times before CallbackClientSide() is invoked, copiedSize will keep the last/upper value, which is ok. What if callback(0) is called after the sequence of calback(sz)? There should be no such situation now: the only one callback(0) is invoked when there is no more data. But I would place some protection above this line, e.g.,
```
if (sz == 0 && !error)
    Must(copiedSize == 0);
```
Do you think this is reasonable?